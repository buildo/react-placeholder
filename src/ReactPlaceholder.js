import React from 'react';
import PropTypes from 'prop-types';
import * as placeholders from './placeholders';

export default class ReactPlaceholder extends React.Component {

  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.node,
      PropTypes.element
    ]).isRequired,
    ready: PropTypes.bool.isRequired,
    firstLaunchOnly: PropTypes.bool,
    type: PropTypes.oneOf(['text', 'media', 'textRow', 'rect', 'round']),
    rows: PropTypes.number,
    color: PropTypes.string,
    showLoadingAnimation: PropTypes.bool,
    customPlaceholder: PropTypes.oneOfType([
      PropTypes.node,
      PropTypes.element
    ])
  }

  static defaultProps = {
    type: 'text',
    color: '#CDCDCD'
  }

  state = {
    ready: this.props.ready
  }

  isReady = () => (
    this.props.firstLaunchOnly ? this.state.ready : this.props.ready
  )

  getFiller = () => {
    const {
      firstLaunchOnly, children, ready, className, // eslint-disable-line no-unused-vars
      type, customPlaceholder, showLoadingAnimation, ...rest
    } = this.props;

    const classes = showLoadingAnimation ?
      ['show-loading-animation', className].filter(c => c).join(' ') :
      className;

    if (customPlaceholder && React.isValidElement(customPlaceholder)) {
      const mergedCustomClasses =
        [customPlaceholder.props.className, classes].filter(c => c).join(' ');
      return React.cloneElement(customPlaceholder,
        { className: mergedCustomClasses });
    } else if (customPlaceholder && !React.isValidElement(customPlaceholder)) {
      throw new Error('ReferenceError',
        'Custom placeholder is not a valid component.');
    }

    const Placeholder = placeholders[type];

    return <Placeholder {...rest} className={classes} />;
  };

  render() {
    return this.isReady() ? this.props.children : this.getFiller();
  }

  componentWillReceiveProps(nextProps) {
    if (!this.state.ready) {
      this.setState({
        ready: nextProps.ready
      });
    }
  }
}
