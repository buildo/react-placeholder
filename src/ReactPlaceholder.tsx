import * as React from 'react';
import * as PropTypes from 'prop-types';
import * as placeholders from './placeholders';

export type CommonProps = {
  children: React.ReactNode,
  /** pass `true` when the content is ready and `false` when it's loading */
  ready: boolean,
  /** delay in millis to wait when passing from ready to NOT ready */
  delay?: number,
  /** if true, the placeholder will never be rendered again once ready becomes true, even if it becomes false again */
  firstLaunchOnly?: boolean,
  className?: string,
  style?: React.CSSProperties
}

export type Props = (CommonProps & {
  /** type of placeholder to use */
  type: 'text' | 'media' | 'textRow' | 'rect' | 'round',
  /** number of rows displayed in 'media' and 'text' placeholders */
  rows?: number,
  /** color of the placeholder */
  color?: string,
  /** pass true to show a nice loading animation on the placeholder */
  showLoadingAnimation?: boolean,
  customPlaceholder?: undefined
}) | (CommonProps & {
  /** pass any renderable content to be used as placeholder instead of the built-in ones */
  customPlaceholder?: React.ReactNode | React.ReactElement<{ [k: string]: any }>,
  type?: undefined,
  rows?: undefined,
  color?: undefined,
  showLoadingAnimation?: undefined
})

export default class ReactPlaceholder extends React.Component<Props> {

  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.node,
      PropTypes.element
    ]).isRequired,
    delay: PropTypes.number,
    ready: PropTypes.bool.isRequired,
    firstLaunchOnly: PropTypes.bool,
    type: PropTypes.oneOf(['text', 'media', 'textRow', 'rect', 'round']),
    rows: PropTypes.number,
    color: PropTypes.string,
    showLoadingAnimation: PropTypes.bool,
    customPlaceholder: PropTypes.oneOfType([
      PropTypes.node,
      PropTypes.element
    ]),
    className: PropTypes.string,
    style: PropTypes.object
  }

  static defaultProps = {
    delay: 0,
    type: 'text',
    color: '#CDCDCD'
  }

  state = {
    ready: this.props.ready
  }

  timeout?: number;

  getFiller = () => {
    const {
      firstLaunchOnly, children, ready, className, // eslint-disable-line no-unused-vars
      type, customPlaceholder, showLoadingAnimation, ...rest
    } = this.props;

    const classes = showLoadingAnimation ?
      ['show-loading-animation', className].filter(c => c).join(' ') :
      className;

    if (customPlaceholder && React.isValidElement(customPlaceholder)) {
      const mergedCustomClasses = [
        customPlaceholder.props.className,
        classes
      ].filter(c => c).join(' ');
      return React.cloneElement(customPlaceholder, { className: mergedCustomClasses });
    } else if (customPlaceholder) {
      return customPlaceholder;
    }

    const Placeholder = placeholders[type!];

    return <Placeholder {...rest} className={classes} />;
  };

  setNotReady = () => {
    const { delay } = this.props;

    if (delay && delay > 0) {
      this.timeout = window.setTimeout(() => {
        if (!this.props.ready) {
          this.setState({ ready: false });
        }
      }, delay);
    } else {
      this.setState({ ready: false });
    }
  }

  setReady = () => {
    if (this.timeout) {
      window.clearTimeout(this.timeout);
    }

    if (!this.state.ready) {
      this.setState({ ready: true });
    }
  }

  render() {
    return this.state.ready ? this.props.children : this.getFiller();
  }

  componentWillReceiveProps(nextProps: Props) {
    if (!this.props.firstLaunchOnly && this.state.ready && !nextProps.ready) {
      this.setNotReady();
    } else if (nextProps.ready) {
      this.setReady();
    }
  }
}
