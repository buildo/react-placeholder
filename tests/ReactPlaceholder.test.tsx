import * as React from 'react';
import { execSync } from 'child_process';
import * as fs from 'fs';
import * as path from 'path';
import ReactPlaceholder from '../src/ReactPlaceholder';
import { shallow } from 'enzyme';
import { act, render } from '@testing-library/react';
import { JSDOM } from "jsdom"
import '@testing-library/jest-dom';
const flatten = require('lodash.flatten');

const dom = new JSDOM()
global.document = dom.window.document
// @ts-ignore
global.window = dom.window

jest.useFakeTimers();

describe('ReactPlaceholder', () => {
  it('renders the text placeholder with 3 rows as the default placeholder', () => {
    const content = <div>Some content still loading...</div>;
    const tree = shallow(
      <ReactPlaceholder ready={false}>
        {content}
      </ReactPlaceholder>
    );
    expect(tree.contains(content)).toBe(false);
    expect(tree.getElements()).toMatchSnapshot();
  });

  it('renders the text placeholder when the content is not ready', () => {
    const content = <div>Some content still loading...</div>;
    const tree = shallow(
      <ReactPlaceholder type="text" rows={2} ready={false}>
        {content}
      </ReactPlaceholder>
    );
    expect(tree.contains(content)).toBe(false);
    expect(tree.getElements()).toMatchSnapshot();
  });

  it('renders the text placeholder with the lineSpacing set', () => {
    const content = <div>Some content still loading...</div>;
    const tree = shallow(
      <ReactPlaceholder type="text" rows={2} ready={false} lineSpacing={3}>
        {content}
      </ReactPlaceholder>
    );
    expect(tree.contains(content)).toBe(false);
    expect(tree.find('TextBlock').prop('lineSpacing')).toBe(3);
    expect(tree.getElements()).toMatchSnapshot();
  });

  it('renders the media placeholder when the content is not ready', () => {
    const content = <div>Some content still loading...</div>;
    const tree = shallow(
      <ReactPlaceholder type="media" rows={2} ready={false}>
        {content}
      </ReactPlaceholder>
    );
    expect(tree.contains(content)).toBe(false);
    expect(tree.getElements()).toMatchSnapshot();
  });

  it('renders the textRow placeholder when the content is not ready', () => {
    const content = <div>Some content still loading...</div>;
    const tree = shallow(
      <ReactPlaceholder type="textRow" ready={false}>
        {content}
      </ReactPlaceholder>
    );
    expect(tree.contains(content)).toBe(false);
    expect(tree.getElements()).toMatchSnapshot();
  });

  it('renders the rect placeholder when the content is not ready', () => {
    const content = <div>Some content still loading...</div>;
    const tree = shallow(
      <ReactPlaceholder type="rect" ready={false}>
        {content}
      </ReactPlaceholder>
    );
    expect(tree.contains(content)).toBe(false);
    expect(tree.getElements()).toMatchSnapshot();
  });

  it('renders the round placeholder when the content is not ready', () => {
    const content = <div>Some content still loading...</div>;
    const tree = shallow(
      <ReactPlaceholder type="round" ready={false}>
        {content}
      </ReactPlaceholder>
    );
    expect(tree.contains(content)).toBe(false);
    expect(tree.getElements()).toMatchSnapshot();
  });

  it('renders the round placeholder with animation when the content is not ready', () => {
    const content = <div>Some content still loading...</div>;
    const tree = shallow(
      <ReactPlaceholder type="round" ready={false} showLoadingAnimation>
        {content}
      </ReactPlaceholder>
    );
    expect(tree.contains(content)).toBe(false);
    expect(tree.getElements()).toMatchSnapshot();
  });

  it('renders a custom placeholder when the content is not ready', () => {
    const content = <div>Some ready content</div>;
    const customPlaceholder = <div>Custom Placeholder</div>;
    const tree = shallow(
      <ReactPlaceholder ready={false} customPlaceholder={customPlaceholder}>
        {content}
      </ReactPlaceholder>
    );
    expect(tree.contains(content)).toBe(false);
    // an empty className is injected to the customer placeholder element,
    // so we'll check the children for now
    expect(tree.text()).toBe(customPlaceholder.props.children);
    expect(tree.getElements()).toMatchSnapshot();
  });

  it('renders the placeholder only after the specified delay', () => {
    const content = <div>Some content still loading...</div>;
    const { container, queryByText, rerender } = render(
      <ReactPlaceholder ready type="text" rows={2} delay={10000}>
        {content}
      </ReactPlaceholder>
    );
    expect(queryByText('Some content still loading...')).toBeInTheDocument();
    rerender(<ReactPlaceholder ready={false} type="text" rows={2} delay={10000}>
      {content}
    </ReactPlaceholder>);
    expect(queryByText('Some content still loading...')).toBeInTheDocument();
    act(() => {
      jest.runAllTimers();
    });
    expect(queryByText('Some content still loading...')).not.toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders the content when it's ready", () => {
    const content = <div>Some ready content</div>;
    const tree = shallow(
      <ReactPlaceholder type="text" rows={2} ready>
        {content}
      </ReactPlaceholder>
    );
    expect(tree.contains(content)).toBe(true);
    expect(tree.getElements()).toMatchSnapshot();
  });

  it("renders content when it's ready, then a placeholder when it's not ready, and finally content again when it's ready", () => {
    const content = <div>Some ready content</div>;
    const { queryByText, container, rerender } = render(
      <ReactPlaceholder type="text" rows={2} ready={true}>
        {content}
      </ReactPlaceholder>
    );
    expect(queryByText('Some ready content')).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();

    rerender(<ReactPlaceholder type="text" rows={2} ready={false}>
      {content}
    </ReactPlaceholder>);

    expect(queryByText('Some ready content')).not.toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();

    rerender(<ReactPlaceholder type="text" rows={2} ready={true}>
      {content}
    </ReactPlaceholder>);

    expect(queryByText('Some ready content')).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders content when firstLaunchOnly is true and ready changes to true, and keeps it rendered when ready changes to false', () => {
    const content = <div>Some ready content</div>;
    const { queryByText, container, rerender } = render(
      <ReactPlaceholder type="rect" ready={false} firstLaunchOnly={true}>
        {content}
      </ReactPlaceholder>
    );
    expect(queryByText('Some ready content')).not.toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();

    rerender(<ReactPlaceholder  type="rect" ready={true} firstLaunchOnly={true}>
      {content}
    </ReactPlaceholder>);

    expect(queryByText('Some ready content')).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();

    rerender(<ReactPlaceholder  type="rect" ready={false} firstLaunchOnly={true}>
      {content}
    </ReactPlaceholder>);

    // content is still rendered
    expect(queryByText('Some ready content')).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('build', () => {
  it('build script generates every needed file', () => {
    execSync('npm run build', { stdio: [] });

    const libPath = path.resolve(__dirname, '../lib');

    const libFiles = fs.readdirSync(libPath);

    const files = flatten(
      libFiles.map(fileInRoot => {
        const filePath = path.resolve(libPath, fileInRoot);
        if (fs.lstatSync(filePath).isDirectory()) {
          return fs
            .readdirSync(filePath)
            .map(fileInFolder => `${fileInRoot}/${fileInFolder}`);
        }
        return fileInRoot;
      })
    );

    expect(files).toMatchSnapshot();
  });
});
