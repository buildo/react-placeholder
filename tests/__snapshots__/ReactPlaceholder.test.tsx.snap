// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactPlaceholder renders a custom placeholder when the content is not ready 1`] = `
[
  <div
    className=""
  >
    Custom Placeholder
  </div>,
]
`;

exports[`ReactPlaceholder renders content when firstLaunchOnly is true and ready changes to true, and keeps it rendered when ready changes to false 1`] = `
<div
  class="rect-shape"
  style="background-color: rgb(205, 205, 205); width: 100%; height: 100%; margin-right: 10px;"
/>
`;

exports[`ReactPlaceholder renders content when firstLaunchOnly is true and ready changes to true, and keeps it rendered when ready changes to false 2`] = `
<div>
  Some ready content
</div>
`;

exports[`ReactPlaceholder renders content when firstLaunchOnly is true and ready changes to true, and keeps it rendered when ready changes to false 3`] = `
<div>
  Some ready content
</div>
`;

exports[`ReactPlaceholder renders content when it's ready, then a placeholder when it's not ready, and finally content again when it's ready 1`] = `
<div>
  Some ready content
</div>
`;

exports[`ReactPlaceholder renders content when it's ready, then a placeholder when it's not ready, and finally content again when it's ready 2`] = `
<div
  class="text-block"
  style="width: 100%;"
>
  <div
    class="text-row"
    style="max-height: 33.333333333333336%; width: 97%; height: 1em; background-color: rgb(205, 205, 205); margin-top: 0px;"
  />
  <div
    class="text-row"
    style="max-height: 33.333333333333336%; width: 100%; height: 1em; background-color: rgb(205, 205, 205); margin-top: 0.7em;"
  />
</div>
`;

exports[`ReactPlaceholder renders content when it's ready, then a placeholder when it's not ready, and finally content again when it's ready 3`] = `
<div>
  Some ready content
</div>
`;

exports[`ReactPlaceholder renders the content when it's ready 1`] = `
[
  <div>
    Some ready content
  </div>,
]
`;

exports[`ReactPlaceholder renders the media placeholder when the content is not ready 1`] = `
[
  <MediaBlock
    color="#CDCDCD"
    rows={2}
  />,
]
`;

exports[`ReactPlaceholder renders the placeholder only after the specified delay 1`] = `
<div
  class="text-block"
  style="width: 100%;"
>
  <div
    class="text-row"
    style="max-height: 33.333333333333336%; width: 97%; height: 1em; background-color: rgb(205, 205, 205); margin-top: 0px;"
  />
  <div
    class="text-row"
    style="max-height: 33.333333333333336%; width: 100%; height: 1em; background-color: rgb(205, 205, 205); margin-top: 0.7em;"
  />
</div>
`;

exports[`ReactPlaceholder renders the rect placeholder when the content is not ready 1`] = `
[
  <RectShape
    color="#CDCDCD"
    rows={3}
  />,
]
`;

exports[`ReactPlaceholder renders the round placeholder when the content is not ready 1`] = `
[
  <RoundShape
    color="#CDCDCD"
    rows={3}
  />,
]
`;

exports[`ReactPlaceholder renders the round placeholder with animation when the content is not ready 1`] = `
[
  <RoundShape
    className="show-loading-animation"
    color="#CDCDCD"
    rows={3}
  />,
]
`;

exports[`ReactPlaceholder renders the text placeholder when the content is not ready 1`] = `
[
  <TextBlock
    color="#CDCDCD"
    rows={2}
  />,
]
`;

exports[`ReactPlaceholder renders the text placeholder with 3 rows as the default placeholder 1`] = `
[
  <TextBlock
    color="#CDCDCD"
    rows={3}
  />,
]
`;

exports[`ReactPlaceholder renders the text placeholder with the lineSpacing set 1`] = `
[
  <TextBlock
    color="#CDCDCD"
    lineSpacing={3}
    rows={2}
  />,
]
`;

exports[`ReactPlaceholder renders the textRow placeholder when the content is not ready 1`] = `
[
  <TextRow
    color="#CDCDCD"
    rows={3}
  />,
]
`;

exports[`build build script generates every needed file 1`] = `
[
  "ReactPlaceholder.d.ts",
  "ReactPlaceholder.js",
  "index.d.ts",
  "index.js",
  "placeholders/MediaBlock.d.ts",
  "placeholders/MediaBlock.js",
  "placeholders/RectShape.d.ts",
  "placeholders/RectShape.js",
  "placeholders/RoundShape.d.ts",
  "placeholders/RoundShape.js",
  "placeholders/TextBlock.d.ts",
  "placeholders/TextBlock.js",
  "placeholders/TextRow.d.ts",
  "placeholders/TextRow.js",
  "placeholders/index.d.ts",
  "placeholders/index.js",
  "reactPlaceholder.css",
  "reactPlaceholder.css.map",
  "utils.d.ts",
  "utils.js",
]
`;
